modules/storage-api/src/main/java/org/apache/ignite/internal/storage/ThreadAssertingMvPartitionStorage.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.storage;
+
+import static org.apache.ignite.internal.worker.ThreadAssertions.assertThreadAllowsToRead;
+import static org.apache.ignite.internal.worker.ThreadAssertions.assertThreadAllowsToWrite;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.schema.BinaryRow;
+import org.apache.ignite.internal.storage.gc.GcEntry;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.internal.worker.ThreadAssertions;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * {@link MvPartitionStorage} that performs thread assertions when doing read/write operations.
+ *
+ * @see ThreadAssertions
+ */
+public class ThreadAssertingMvPartitionStorage implements MvPartitionStorage {
```

Comment:
Can you please explain the reason of having a wrapper class instead of putting assertions into `SnapshotAwarePartitionDataStorage`, for example? Or maybe we should put some assertions in places where we use API, not into the implementation.
Let me elaborate. For example, what if we have a cursor and somebody reads from it? Should we create wrapper for it as well? Should we create wrapper for all cursors in storages?
It makes the code too complicated in my opinion, too many classes and too many wrappers, and what are the benefits? The fact that you would place assertion into a single implementation rather than several implementations for several engines (potentially)? This is arguable. All I see is a code that re-implements a lot of methods of existing interface, forcing us to modify more classes if we modify the interface itself.